pkg/operator/ceph/object/notification/obc_label_controller.go

```diff
@@ -149,37 +152,60 @@ func (r *ReconcileOBCLabels) reconcile(request reconcile.Request) (reconcile.Res
 				logger.Warningf("bucket notification label mismatch. ignoring key %q value %q", labelKey, labelValue)
 				continue
 			}
-			// for each notification label fetch the bucket notification CRD
+			labelList = append(labelList, labelValue)
 			logger.Debugf("bucket notification label %q found on ObjectbucketClaim %q", labelValue, bucketName)
-			notification := &cephv1.CephBucketNotification{}
-			bnName := types.NamespacedName{Namespace: obc.Namespace, Name: labelValue}
-			if err := r.client.Get(r.opManagerContext, bnName, notification); err != nil {
-				if kerrors.IsNotFound(err) {
-					logger.Infof("CephBucketNotification %q not found", bnName)
-					return waitForRequeueIfNotificationNotReady, nil
-				}
-				return reconcile.Result{}, errors.Wrapf(err, "failed to retrieve CephBucketNotification %q", bnName)
-			}
+		}
+	}
 
-			// get the topic associated with the notification, and make sure it is provisioned
-			topicName := types.NamespacedName{Namespace: notification.Namespace, Name: notification.Spec.Topic}
-			bucketTopic, err := topic.GetProvisioned(r.client, r.opManagerContext, topicName)
-			if err != nil {
-				logger.Infof("CephBucketTopic %q not provisioned yet", topicName)
-				return waitForRequeueIfTopicNotReady, nil
-			}
+	// remove notifications which are no longer specified in the OBC labels
+	for _, oldValue := range bnList {
+		if !sets.NewString(labelList...).Has(oldValue) {
+			deleteList = append(deleteList, oldValue)
+		}
+	}
+	for _, notificationId := range deleteList {
+		err = deleteNotificationFunc(p, &ob, notificationId)
+		if err != nil {
+			logger.Warningf("notification %q failed remove from %q, returned error %v", notificationId, ob.Spec.Endpoint.BucketName, err)
+		}
+	}
 
-			if err = validateObjectStoreName(bucketTopic, objectStoreName); err != nil {
-				return reconcile.Result{}, err
-			}
+	// add new notifications to the list
+	return r.addNewNotifications(p, ob, labelList, objectStoreName, obc.Namespace)
+}
 
-			// provision the notification
-			err = createNotificationFunc(p, &ob, *bucketTopic.Status.ARN, notification)
-			if err != nil {
-				return reconcile.Result{}, errors.Wrapf(err, "failed to provision CephBucketNotification %q", bnName)
+func (r *ReconcileOBCLabels) addNewNotifications(p provisioner, ob bktv1alpha1.ObjectBucket, notificationList []string, objectStoreName types.NamespacedName, namespace string) (reconcile.Result, error) {
+	for _, labelValue := range notificationList {
+		logger.Debugf("adding bucket notification %q", labelValue)
```

Comment:
Missing some contextual information about the source of config info. This should identify the notification by namespace and name, otherwise it could be hard from context to verify if the notification is right/wrong in the event of a bug. We should also identify the OBC name as well.